### DISCUSSION TITLE
replace the embedded font by just replacing the whole font contents directly

### DISCUSSION THREAD

**Discussion by JackieFei** [pymupdf/PyMuPDF#918] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**

for example : pdf already embedded full set of `serif.ttf`, wish to be replaced by full set of `sans.ttf`

Can we replace the embedded font by just replacing the whole font contents directly? (without re-writing text)

reference:
https://github.com/pymupdf/PyMuPDF/discussions/855
https://github.com/pymupdf/PyMuPDF-Utilities/tree/master/font-replacement
(The repl-font.py is using TextWriter to re-write text)

**Describe the solution you'd like**

flow description: 
1. use pymupdf fitz open pdf
2. get `serif.ttf` xref number
3. use some methods to get `sans.ttf` object contents
4. use some methods (+xref number) to replace `serif.ttf` object contents with `sans.ttf` object contents
5. save (`sans.ttf` embedded, `serif.ttf` dropped) new pdf

expect: 
pdf all text written by `serif.ttf` can displayed with `sans.ttf`

**Describe alternatives you've considered**

**Additional context**

**Comment by JorjMcKie** [comment, marked as answer]
No, this is impossible - meaning it **_cannot ever_** work. Not even if you wanted to replace a sans font by some other sans font.

There are several reasons for this, here are just two:
* every single glyph has its own metrics (at least for proportional fonts), e.g. an "a" in Palatino has a different width than in Arial. In addition the **_relative_** width within a font is different from that of another font.
* Text in a PDF is encoded with a **_glyph number_** per unicode. This number is in general not equal to the unicode number and it is font-specific. So if text is not rewritten with the new font, the result would be unreadable.

The next version will however contain a new `Document.subset_fonts()` that replaces a font with its subset - without rewriting text.

**Reply by JorjMcKie** [reply]
Example: 3 sans fonts. Closely look at the word "Hello" and you will notice different absolute and relative, intra-font character widths:
![grafik](https://user-images.githubusercontent.com/8290722/108871597-b1681300-75cf-11eb-8d0b-c8bf9b829005.png)
All three use a different mapping of unicode -> glyph number:
```
BT
/F0 11 Tf
1 0 0 1 100 742 Tm
[<00290046004D004D0050000D000100290046004D005700460055004A004400420002>] TJ
0 G
/F1 11 Tf
1 0 0 1 100 722 Tm
[<002B0048004F004F0052000F0003003100520057005200030036004400510056000300350048004A0058004F004400550004>] TJ
0 G
/F2 11 Tf
1 0 0 1 100 702 Tm
[<007301A9021B021B02440E850003006501F6027C015A006800B9000300EC01A901D702AF021B015A027C0E89>] TJ
ET
```
The first 4 hex digits for example in each case represent the glyph number for "H" of "Hello". So 0x0029 = "H" in Helvetica, 0x002B in Noto Sans and 0x0073 in Fira Go.
Correspondingly, if I now do a cyclic exchange of the 3 fonts (`/F0`->`/F1`->`/F2`->`/F0`) the PDF will show you this:
![grafik](https://user-images.githubusercontent.com/8290722/108872638-bda0a000-75d0-11eb-9cd9-3852c0b4405e.png)

**Reply by JackieFei** [reply]
Thanks for your test.

Q1: Can you make a more test with the same company's fonts, just like `Noto Serif` and `Noto Sans`, maybe the text _glyph numbers_ are identical?

Q2 : Is the action `do a cyclic exchange of the 3 fonts` done by PyMuPDF functions? How to do it?

Thanks in advanced.

**Reply by JorjMcKie** [reply]
Q1: Sometimes they are, sometimes they are not. You can check that yourself: `Font.has_glyph(ord("A"))` returns the glyph number.
Q2: I just used an editor and changed the numbers manually.

**Reply by JorjMcKie** [reply]
Addendum for Q1: for a font supported by MuPDF, there is the array `Font.valid_codepoints()` which delivers the unicode number for every existing glyph id of the font. Two fonts have the same mapping if these arrays are equal obviously.
E.g. `fitz.Font("tiro").valid_codepoints() == fitz.Font("helv).valid_codepoints()` is `True`.

**Reply by JorjMcKie** [reply]
In any case, my first objection remains true: `fitz.Font("helv").text_length(text) != fitz.Font("tiro").text_length(text)` practically always - even though these two (and "cour" too) happen to have the same glyph mapping.

**Comment by JorjMcKie** [comment]
A final comment on this:
Version 1.18.9 will have an additional way to insert a font in a PDF. Then you can experiment a little by doing this kind of thing:
```python
# assume 'xref1' is the xref of a font you wish to replace
# insert the desired new font:
xref2 = doc._insert_font(filename=...)
str_xref2 = doc.xref_object(xref2)  # get definition syntax of new font
doc.update_object(xref1, str_xref2)  # replace old font defintion
...
doc.save(..., garbage=3,...)  # will remove unused objects
```
